{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial: Integrating GraphSpace into network analysis projects\n",
    "See session2.ipynb to get set up.\n",
    "\n",
    "About Me:\n",
    "  - Jeff Law\n",
    "  - PhD Student, Genetics, Bioinformatics and Computational Biology at Virginia Tech\n",
    "  - Email: jeffl@vt.edu\n",
    "\n",
    "## Agenda\n",
    "1. Pipeline Overview\n",
    "2. Introduction to PathLinker\n",
    "  - Protein-protein interaction network\n",
    "  - k shortest paths\n",
    "3. Post results to GraphSpace\n",
    "  - Read in PathLinker results\n",
    "  - Add style, popups, path rank filter and more attributes\n",
    "4. Reconstruction of the WNT Signaling Pathway\n",
    "5. Post your own networks!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Introduction to PathLinker\n",
    "PathLinker is an implementation of the k shortest paths algorith known as [Yen's Algorithm](https://en.wikipedia.org/wiki/Yen%27s_algorithm).\n",
    "PathLinker has some useful additional features\n",
    "  - shortest paths from multiple sources to multiple targets\n",
    "\n",
    "<p><a href=\"https://commons.wikimedia.org/wiki/File:Yen%27s_K-Shortest_Path_Algorithm,_K%3D3,_A_to_F.gif#/media/File:Yen%27s_K-Shortest_Path_Algorithm,_K%3D3,_A_to_F.gif\"><img src=\"https://upload.wikimedia.org/wikipedia/commons/d/d1/Yen%27s_K-Shortest_Path_Algorithm%2C_K%3D3%2C_A_to_F.gif\" alt=\"Yen's k-shortest path algorithm, K&nbsp;=&nbsp;3, A to F\"></a><br>By <a href=\"//commons.wikimedia.org/w/index.php?title=User:KRPent&amp;action=edit&amp;redlink=1\" class=\"new\" title=\"User:KRPent (page does not exist)\">KRPent</a> - Using an open source graphing library, Graphviz, to generate the image., <a href=\"http://creativecommons.org/licenses/by-sa/3.0\" title=\"Creative Commons Attribution-Share Alike 3.0\">CC BY-SA 3.0</a>, <a href=\"https://commons.wikimedia.org/w/index.php?curid=22926392\">Link</a></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run PathLinker\n",
    "First we need to download PathLinker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'PathLinker'...\n",
      "remote: Counting objects: 168, done.\u001b[K\n",
      "remote: Total 168 (delta 0), reused 0 (delta 0), pack-reused 168\u001b[K\n",
      "Receiving objects: 100% (168/168), 2.68 MiB | 0 bytes/s, done.\n",
      "Resolving deltas: 100% (72/72), done.\n",
      "Checking connectivity... done.\n"
     ]
    }
   ],
   "source": [
    "!git clone \"https://github.com/Murali-group/PathLinker\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein-protein interaction network\n",
    "Protein-protein interactions (PPIs) are available from multiple databases for many species. \n",
    "We work with Human and Yeast \n",
    "\n",
    "Sources: BioGrid, DIP, InnateDB, IntAct, KEGG, MatrixDB, MINT, NetPath, PhosphositePlus, Reactome, SPIKE'\n",
    "  - 35K regulatory (directed) PPIs\n",
    "  - 200K physical (undirected) PPIs\n",
    "  \n",
    "Edges are weighted using an evidence-based probabilistic approach.\n",
    "\n",
    "### Sources and Targets\n",
    "Using the WNT pathway receptors as sources and the transcription facotrs (TFs) as targets, we can use PathLinker to build a reconstruction of the WNT pathway.\n",
    "\n",
    "### Running PathLinker\n",
    "Now that we have all of the inputs set, we can run PathLinker\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \r\n",
      "PathLinker.py [options] NETWORK NODE_TYPES\r\n",
      "REQUIRED arguments:\r\n",
      "    NETWORK - A tab-delimited file with one directed interaction per\r\n",
      "        line. Each line should have at least 2 columns: tail, head. Edges\r\n",
      "        are directed from tail to head. This file can have a third column\r\n",
      "        specifying the edge weight, which is required unless the --PageRank\r\n",
      "        option is used (see --PageRank help for a note on these weights).\r\n",
      "        To run PathLinker on an unweighted graph, set all edge weights\r\n",
      "        to 1 in the input network.\r\n",
      "\r\n",
      "    NODE_TYPES - A tab-delimited file denoting nodes as receptors or TRs. The first\r\n",
      "        column is the node name, the second is the node type, either 'source'\r\n",
      "        (or 'receptor') or 'target' (or 'tr' or 'tf'). Nodes which are neither receptors nor TRs may\r\n",
      "        be omitted from this file or may be given a type which is neither 'source'\r\n",
      "        nor 'target'.\r\n",
      "    \r\n",
      "\r\n",
      "\r\n",
      "Options:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "  -o STR, --output=STR  A string to prepend to all output files.\r\n",
      "                        (default=\"out\")\r\n",
      "  --write-paths         If given, also output a list of paths found by KSP in\r\n",
      "                        addition to the ranked edges.\r\n",
      "  --no-log-transform    Normally input edge weights are log-transformed. This\r\n",
      "                        option disables that step.\r\n",
      "  --largest-connected-component\r\n",
      "                        Run PathLinker on only the largest weakly connected\r\n",
      "                        component of the graph. May provide performance\r\n",
      "                        speedup.\r\n",
      "  --edge-penalty=EDGE_PENALTY\r\n",
      "                        Factor by which to divide every edge weight. The\r\n",
      "                        effect of this option is to penalize the score of\r\n",
      "                        every path by a factor equal to (the number of edges\r\n",
      "                        in the path)^(this factor). (default=1.0)\r\n",
      "\r\n",
      "  Random Walk Options:\r\n",
      "    --PageRank          Run the PageRank algorithm to generate edge visitation\r\n",
      "                        flux values, which are then used as weights for KSP. A\r\n",
      "                        weight column in the network file is not needed if\r\n",
      "                        this option is given, as the PageRank visitation\r\n",
      "                        fluxes are used for edge weights in KSP. If a weight\r\n",
      "                        column is given, these weights are interpreted as a\r\n",
      "                        weighted PageRank graph.\r\n",
      "    -q Q_PARAM, --q-param=Q_PARAM\r\n",
      "                        The value of q indicates the probability that the\r\n",
      "                        random walker teleports back to a source node during\r\n",
      "                        the random walk process. (default=0.5)\r\n",
      "    -e EPSILON, --epsilon=EPSILON\r\n",
      "                        A small value used to test for convergence of the\r\n",
      "                        iterative implementation of PageRank. (default=0.0001)\r\n",
      "    --max-iters=MAX_ITERS\r\n",
      "                        Maximum number of iterations to run the PageRank\r\n",
      "                        algorithm. (default=500)\r\n",
      "\r\n",
      "  k Shortest Paths Options:\r\n",
      "    -k K_PARAM, --k-param=K_PARAM\r\n",
      "                        The number of shortest paths to find. (default=100)\r\n",
      "    --allow-mult-targets\r\n",
      "                        By default, PathLinker will remove outgoing edges from\r\n",
      "                        targets to ensure that there is only one target on\r\n",
      "                        each path.  If --allow-mult-targets is specified,\r\n",
      "                        these edges are not removed.\r\n",
      "    --allow-mult-sources\r\n",
      "                        By default, PathLinker will remove incoming edges to\r\n",
      "                        sources to ensure that there is only one source on\r\n",
      "                        each path.  If --allow-mult-sources is specified,\r\n",
      "                        these edges are not removed.\r\n"
     ]
    }
   ],
   "source": [
    "!python PathLinker/PathLinker.py --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading the network from data/2017-06-16-human-ppi-weighted-cap0_75.tsv\n",
      "Name: \n",
      "Type: DiGraph\n",
      "Number of nodes: 16421\n",
      "Number of edges: 438775\n",
      "Average in degree:  26.7204\n",
      "Average out degree:  26.7204\n",
      "Reading sources and targets from data/Wnt-sources-targets.tsv\n",
      "\n",
      "Read 14 sources and 14 targets\n",
      "\tAfter removing sources and targets that are not in the network: 14 sources and 14 targets.\n",
      "\n",
      "Computing the k=100 shortest simple paths.\n",
      "\n",
      "KSP results are in \"output/Wnt-k_100-ranked-edges.txt\"\n",
      "KSP paths are in \"output/Wnt-k_100-paths.txt\"\n",
      "\n",
      "Finished!\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir -p output\n",
    "python PathLinker/PathLinker.py \\\n",
    "    data/2017-06-16-human-ppi-weighted-cap0_75.tsv \\\n",
    "    data/Wnt-sources-targets.tsv \\\n",
    "    -k 100 \\\n",
    "    -o output/Wnt- \\\n",
    "    --write-paths "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> output/Wnt-k_100-paths.txt <==\r\n",
      "#KSP\tpath_length\tpath\r\n",
      "1\t7.50000e-01\tO75581|P98082\r\n",
      "2\t5.62500e-01\tO00144|O14640|P25054\r\n",
      "3\t5.62500e-01\tO00144|O14640|P35222\r\n",
      "4\t5.62500e-01\tO00144|O14640|Q15797\r\n",
      "5\t5.62500e-01\tO00144|O14641|P25054\r\n",
      "6\t5.62500e-01\tO00144|Q92997|P98082\r\n",
      "7\t5.62500e-01\tO75084|O14640|P25054\r\n",
      "8\t5.62500e-01\tO75084|O14640|P35222\r\n",
      "9\t5.62500e-01\tO75084|O14640|Q15797\r\n",
      "\r\n",
      "==> output <==\r\n",
      "head: error reading 'output': Is a directory\r\n",
      "\r\n",
      "==> output/Wnt-k_100-ranked-edges.txt <==\r\n",
      "#tail\thead\tKSP index\r\n",
      "O75581\tP98082\t1\r\n",
      "O00144\tO14640\t2\r\n",
      "O14640\tP25054\t2\r\n",
      "O14640\tP35222\t3\r\n",
      "O14640\tQ15797\t4\r\n",
      "O00144\tO14641\t5\r\n",
      "O14641\tP25054\t5\r\n",
      "O00144\tQ92997\t6\r\n",
      "Q92997\tP98082\t6\r\n"
     ]
    }
   ],
   "source": [
    "!head output/Wnt-k_100-paths.txt output output/Wnt-k_100-ranked-edges.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Post results to GraphSpace\n",
    "This step is not too difficult if we just want to view the nodes, edges and paths on GraphSpace. However, there is much more information we have that we would like to be able to quickly access and see on GraphSpace. For example:\n",
    "  - What sources of evidence do we have for each interaction in the top 100 paths?\n",
    "  - Which proteins and PPIs in the top 100 paths are known to be part of the NetPath or KEGG WNT pathway?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: graphspace_python in /home/jeffl/git-workspace/src/graphspace-python\n",
      "Requirement already satisfied: future in /home/jeffl/.local/lib/python2.7/site-packages (from graphspace_python)\n",
      "Requirement already satisfied: networkx==1.11 in /home/jeffl/.local/lib/python2.7/site-packages (from graphspace_python)\n",
      "Requirement already satisfied: requests==2.10.0 in /home/jeffl/.local/lib/python2.7/site-packages (from graphspace_python)\n",
      "Requirement already satisfied: sphinx-rtd-theme in /home/jeffl/.local/lib/python2.7/site-packages (from graphspace_python)\n",
      "Requirement already satisfied: sphinx in /home/jeffl/.local/lib/python2.7/site-packages (from graphspace_python)\n",
      "Requirement already satisfied: recommonmark in /home/jeffl/.local/lib/python2.7/site-packages (from graphspace_python)\n",
      "Requirement already satisfied: decorator>=3.4.0 in /home/jeffl/.local/lib/python2.7/site-packages (from networkx==1.11->graphspace_python)\n",
      "Requirement already satisfied: sphinxcontrib-websupport in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: six>=1.5 in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: docutils>=0.11 in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: typing; python_version < \"3.5\" in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: Pygments>=2.0 in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: setuptools in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: imagesize in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: babel!=2.0,>=1.3 in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: snowballstemmer>=1.1 in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: Jinja2>=2.3 in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: alabaster<0.8,>=0.7 in /home/jeffl/.local/lib/python2.7/site-packages (from sphinx->graphspace_python)\n",
      "Requirement already satisfied: commonmark<=0.5.4 in /home/jeffl/.local/lib/python2.7/site-packages (from recommonmark->graphspace_python)\n",
      "Requirement already satisfied: pytz>=0a in /home/jeffl/.local/lib/python2.7/site-packages (from babel!=2.0,>=1.3->sphinx->graphspace_python)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /home/jeffl/.local/lib/python2.7/site-packages (from Jinja2>=2.3->sphinx->graphspace_python)\n"
     ]
    }
   ],
   "source": [
    "# Make sure graphspace-python library is installed\n",
    "!pip install graphspace_python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "94 edges, 14 sources, 14 targets\n"
     ]
    }
   ],
   "source": [
    "# Read in the PathLinker results and keep track of the k for each of the edges\n",
    "ranked_edges = {}\n",
    "with open('output/Wnt-k_100-ranked-edges.txt', 'r') as file_handle:\n",
    "    for line in file_handle:\n",
    "        if line[0] == '#':\n",
    "            continue\n",
    "        line = line.rstrip().split('\\t')\n",
    "        edge = (line[0], line[1])\n",
    "        k = line[2]\n",
    "        ranked_edges[edge] = k\n",
    "        \n",
    "# Read in the sources and targets file so we can change the shape and color of the sources and targets\n",
    "sources = set()\n",
    "targets = set()\n",
    "with open('data/Wnt-sources-targets.tsv', 'r') as file_handle:\n",
    "    for line in file_handle:\n",
    "        if line[0] == '#':\n",
    "            continue\n",
    "        line = line.rstrip().split('\\t')\n",
    "        prot = line[0]\n",
    "        prot_type = line[1]\n",
    "        if prot_type.lower() in [\"receptor\", \"source\"]:\n",
    "            sources.add(prot)\n",
    "        elif prot_type.lower() in [\"tf\", \"target\"]:\n",
    "            targets.add(prot)\n",
    "\n",
    "print(\"%d edges, %d sources, %d targets\" % (len(ranked_edges), len(sources), len(targets)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "438775 interactome edges\n",
      "132 kegg wnt nodes, 940 kegg wnt edges\n",
      "81 netpath wnt nodes, 245 netpath wnt edges\n"
     ]
    }
   ],
   "source": [
    "# before adding these edges to a GraphSpace object, \n",
    "# we need to get the node, edge, and pathway information we want to post\n",
    "# we also need to be able to map the uniprot IDs to their more friendly gene names\n",
    "uniprot_to_gene = {}\n",
    "with open('data/2017_06-human-interactome-mapping.tsv', 'r') as file_handle:\n",
    "    for line in file_handle:\n",
    "        if line[0] == '#':\n",
    "            continue\n",
    "        uniprotID, gene_name = line.rstrip().split('\\t')\n",
    "        uniprot_to_gene[uniprotID] = gene_name\n",
    "        \n",
    "# get the edge weights from the interactome\n",
    "edge_weights = {}\n",
    "with open('data/2017-06-16-human-ppi-weighted-cap0_75.tsv', 'r') as file_handle:\n",
    "    for line in file_handle:\n",
    "        if line[0] == '#':\n",
    "            continue\n",
    "        u, v, w, evidence = line.rstrip().split('\\t')\n",
    "        edge_weights[(u,v)] = float(w)\n",
    "        \n",
    "print(\"%s interactome edges\" % (len(edge_weights)))\n",
    "        \n",
    "# get the netpath and kegg edges and nodes\n",
    "# KEGG WNT pathway: http://www.genome.jp/kegg-bin/show_pathway?hsa04310\n",
    "# NetPath WNT pathway: http://www.netpath.org/pathways?path_id=NetPath_8\n",
    "kegg_wnt_edges = set()\n",
    "netpath_wnt_edges = set()\n",
    "evidence_lines = 'data/2017_06-human-interactome-evidence.tsv'\n",
    "with open(evidence_lines, 'r') as file_handle:\n",
    "    for line in file_handle:\n",
    "        if line[0] == \"#\":\n",
    "            continue\n",
    "        u,v, directed, interactiontype, detectionmethod, pubid, source = line.rstrip().split('\\t')\n",
    "        if source.lower() == 'kegg' and pubid.lower() == 'kegg:hsa04310':\n",
    "            kegg_wnt_edges.add((u,v))\n",
    "            if directed == \"False\":\n",
    "                kegg_wnt_edges.add((v,u))\n",
    "        elif source.lower() == 'netpath' and pubid.lower() == \"netpath:netpath_8\":\n",
    "            netpath_wnt_edges.add((u,v))\n",
    "            if directed == \"False\":\n",
    "                netpath_wnt_edges.add((v,u))\n",
    "kegg_wnt_nodes = set([n for u,v in kegg_wnt_edges for n in (u,v)])\n",
    "netpath_wnt_nodes = set([n for u,v in netpath_wnt_edges for n in (u,v)])\n",
    "\n",
    "print(\"%s kegg wnt nodes, %s kegg wnt edges\" % (len(kegg_wnt_nodes), len(kegg_wnt_edges)))\n",
    "print(\"%s netpath wnt nodes, %s netpath wnt edges\" % (len(netpath_wnt_nodes), len(netpath_wnt_edges)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# this next section is to get the sources of evidence, interaction type, and pubmed IDs for each interaction\n",
    "# this is a bit more difficult, but we can load it into a multi-level dictionary\n",
    "def getEvidence(edges, evidence_file):\n",
    "    \"\"\"\n",
    "    *edges*: a set of edges for which to get the evidence for\n",
    "    *split_family_nodes*: add evidence of family edges to pairwise interactions \n",
    "    *add_ev_to_family_edges*: add evidence of PPI edge covered by a family edge to the family edge. \n",
    "                              If this option is specified, then a dictionary of direct evidence along with\n",
    "                              a dictionary of the ppi evidence applied to the family edge is returned\n",
    "    returns a multi-level dictionary with the following structure\n",
    "    edge: \n",
    "      db/source: \n",
    "        interaction_type: \n",
    "          detection_method: \n",
    "            publication / database ids\n",
    "    for NetPath, KEGG and SPIKE, the detection method is the pathway name \n",
    "    NetPath, KEGG, and Phosphosite also have a pathway ID in the database/id set\n",
    "    which follows convention of id_type:id (for example: kegg:hsa04611)\n",
    "    \"\"\"\n",
    "    # initialize the evidence, edge_types, and edge dir/undir dictionaries\n",
    "    evidence = {}\n",
    "    edge_types = {}\n",
    "    edge_dir = {}\n",
    "\n",
    "    print \"Reading evidence file %s\" % (evidence_file)\n",
    "\n",
    "    # create a graph of the passed in edges \n",
    "    G = nx.Graph()\n",
    "    G.add_edges_from(edges)\n",
    "\n",
    "    # initialize the dictionaries\n",
    "    for t,h in G.edges():\n",
    "        evidence[(t,h)] = {}\n",
    "        evidence[(h,t)] = {}\n",
    "        edge_types[(t,h)] = set()\n",
    "        edge_types[(h,t)] = set()\n",
    "        edge_dir[(t,h)] = False\n",
    "        edge_dir[(h,t)] = False\n",
    "\n",
    "    file_handle = open(evidence_lines, 'r')\n",
    "    for line in file_handle:\n",
    "        if line[0] == \"#\":\n",
    "            continue\n",
    "        u,v, directed, interactiontype, detectionmethod, pubid, source = line.rstrip().split('\\t')\n",
    "        # header line of file\n",
    "        #uniprot_a  uniprot_b   directed    interaction_type    detection_method    publication_id  source\n",
    "        directed = True if directed == \"True\" else False\n",
    "\n",
    "        # We only need to get the evidence for the edges passed in, so if this edge is not in the list of edges, skip it\n",
    "        # G is a Graph so it handles undirected edges correctly\n",
    "        if not G.has_edge(u,v):\n",
    "            continue\n",
    "\n",
    "        evidence = addToEvidenceDict(evidence, (u,v), directed, source, interactiontype, detectionmethod, pubid)\n",
    "        edge_types = addEdgeType(edge_types, (u,v), directed, source, interactiontype)\n",
    "\n",
    "        if directed is True:\n",
    "            edge_dir[(u,v)] = True\n",
    "        else:\n",
    "            # if they are not already directed, then set them as undirected\n",
    "            if (u,v) not in edge_dir:\n",
    "                edge_dir[(u,v)] = False\n",
    "                edge_dir[(v,u)] = False\n",
    "\n",
    "    return evidence, edge_types, edge_dir\n",
    "\n",
    "\n",
    "    \n",
    "def addToEvidenceDict(evidence, e, directed, source, interactiontype, detectionmethod, pubid):\n",
    "    \"\"\" add the evidence of the edge to the evidence dictionary\n",
    "    *pubids*: publication id to add to this edge. \n",
    "    \"\"\"\n",
    "    if source not in evidence[e]:\n",
    "        evidence[e][source] = {}\n",
    "    if interactiontype not in evidence[e][source]:\n",
    "        evidence[e][source][interactiontype] = {}\n",
    "    if detectionmethod not in evidence[e][source][interactiontype]:\n",
    "        evidence[e][source][interactiontype][detectionmethod] = set()\n",
    "    evidence[e][source][interactiontype][detectionmethod].add(pubid)\n",
    "\n",
    "    if not directed:\n",
    "        # add the evidence for both directions\n",
    "        evidence = addToEvidenceDict(evidence, (e[1],e[0]), True, source, interactiontype, detectionmethod, pubid)\n",
    "\n",
    "    return evidence\n",
    "\n",
    "\n",
    "def addEdgeType(edge_types, e, directed, source, interactiontype):\n",
    "    if e not in edge_types:\n",
    "        edge_types[e] = set()\n",
    "    # add the edge type as well\n",
    "    # direction was determined using the csbdb_interface.psimi_interaction_direction dictionary in CSBDB. \n",
    "    if not directed:\n",
    "        # it would be awesome if we knew which edges are part of complex formation vs other physical interactions\n",
    "        # is there an mi term for that?\n",
    "        t,h = e\n",
    "        edge_types[(t,h)].add('physical')\n",
    "        edge_types[(h,t)].add('physical')\n",
    "\n",
    "    elif source == \"SPIKE\":\n",
    "        edge_types[e].add('spike_regulation')\n",
    "\n",
    "    elif source == \"KEGG\":\n",
    "        if \"phosphorylation\" in interactiontype and \"dephosphorylation\" not in interactiontype:\n",
    "            # Most of them are phosphorylation\n",
    "            edge_types[e].add('phosphorylation')\n",
    "        if \"activation\" in interactiontype:\n",
    "            edge_types[e].add('activation')\n",
    "        if \"inhibition\" in interactiontype:\n",
    "            edge_types[e].add('inhibition')\n",
    "        else:\n",
    "            edge_types[e].add('enzymatic')\n",
    "    else:\n",
    "        # the rest is psi mi tags\n",
    "        # MI:0217 is for phosphorylation\n",
    "        if \"MI:0217\" in interactiontype:\n",
    "            # Most of the directed edges are phosphorylation\n",
    "            edge_types[e].add('phosphorylation')\n",
    "        else:\n",
    "            # TODO for now, just call the rest of the directed edges enzymatic. \n",
    "            edge_types[e].add('enzymatic')\n",
    "\n",
    "    return edge_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# converts the evidence dictionary to html for the popup\n",
    "def evidenceToHTML(u,v,evidence):\n",
    "    annotation = '<dl>'\n",
    "    sources = sorted(evidence.keys())\n",
    "    for source in sources:\n",
    "        annotation += '<dt>%s</dt>' % (source)\n",
    "        # TODO add interaction type color\n",
    "        for interactiontype in evidence[source]:\n",
    "            if interactiontype != '' and interactiontype != \"None\":\n",
    "                # use bull instead of li to save on white space\n",
    "                # nbsp stands for non-breaking space\n",
    "                annotation += '&bull;&nbsp&nbsp%s <br>' % interactiontype\n",
    "            for detectionmethod in evidence[source][interactiontype]:\n",
    "                # add a bullet point here after 4 spaces\n",
    "                annotation += '&nbsp&nbsp&nbsp&nbsp&bull;&nbsp&nbsp'\n",
    "                annotation += '%s  ' % detectionmethod\n",
    "\n",
    "                # now add the pubmed IDs. &nbsp is the html for a non-breaking space\n",
    "                pub_ids = evidence[source][interactiontype][detectionmethod]\n",
    "                #KEGG doesn't have pub ids. It has a pathway map and entry (evidence)\n",
    "                # now get the html for each of the links\n",
    "                pub_ids = [parsePubID(pub_id) for pub_id in pub_ids if parsePubID(pub_id) != '']\n",
    "\n",
    "                # use a non-breaking space with a comma so they all stay on the same line\n",
    "                annotation += ',&nbsp'.join(pub_ids)\n",
    "                annotation += \"<br>\"\n",
    "        annotation += '<br>'\n",
    "    return annotation\n",
    "\n",
    "# adds the links to the pubmed and other IDs\n",
    "def parsePubID(publication_id):\n",
    "    id_type, pubid = publication_id.split(':')\n",
    "    if id_type == 'pubmed':\n",
    "        pubmedurl = 'http://www.ncbi.nlm.nih.gov/pubmed/%s' % (pubid)\n",
    "        desc = '<a style=\"color:blue\" href=\"%s\" target=\"PubMed\">pmid:%s</a>' % (pubmedurl,pubid)\n",
    "    elif id_type == 'phosphosite':\n",
    "        phosphourl = 'http://www.phosphosite.org/siteAction.action?id=%s' % (pubid)\n",
    "        desc = '<a style=\"color:blue\" href=\"%s\" target=\"PSP\">phosphosite:%s</a>' % (phosphourl,pubid)\n",
    "    elif id_type == 'kegg':\n",
    "        # links to KEGG pathway map\n",
    "        kegg_map_link = 'http://www.kegg.jp/kegg-bin/show_pathway?'\n",
    "        # links to KEGG pathway entry (evidence)\n",
    "        kegg_entry_link = 'http://www.kegg.jp/dbget-bin/www_bget?pathway+'\n",
    "        pathway_map_link = '<a style=\"color:blue\" href=\"%s%s\" target=\"KEGG\">map</a>' % (kegg_map_link,pubid)\n",
    "        pathway_entry_link = '<a style=\"color:blue\" href=\"%s%s\" target=\"KEGG\">evidence</a>' % (kegg_entry_link,pubid)\n",
    "        desc = \"%s,&nbsp%s\"%(pathway_map_link,pathway_entry_link)\n",
    "    elif id_type == 'netpath':\n",
    "        netpath_url = \"http://www.netpath.org/reactions?path_id=%s\" % (pubid)\n",
    "        # links to KEGG pathway entry (evidence)\n",
    "        desc = '<a style=\"color:blue\" href=\"%s\" target=\"NetPath\">netpath:%s</a>' % (netpath_url, pubid)\n",
    "    else:\n",
    "        # skip the rest for now\n",
    "        desc = ''\n",
    "\n",
    "    return desc\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# takes in a node ID and the list of paths its in, and returns the html popup\n",
    "def getNodePopup(n, pathswithnode):\n",
    "    htmlstring = ''\n",
    "    uniproturl = 'http://www.uniprot.org/uniprot/%s' \n",
    "    htmlstring += '<b>Uniprot ID</b>: <a style=\"color:blue\" href=\"%s\" target=\"UniProtKB\">%s</a><br>' % (uniproturl%n, n)\n",
    "    htmlstring += '<hr />'\n",
    "\n",
    "    htmlstring += '<b>Paths</b>: %s<br>' %(','.join(str(k) for k in sorted(pathswithnode)))\n",
    "    \n",
    "    return htmlstring\n",
    "\n",
    "\n",
    "# takes in the u->v edge, first path index k its in, and sources of evidence \n",
    "# returns the html edge popup\n",
    "def getEdgePopup(u, v, k, evidence):\n",
    "    annotation = '' \n",
    "    annotation += '<b>%s - %s</b></br>'%(uniprot_to_gene[u], uniprot_to_gene[v])\n",
    "    annotation += '<b>%s - %s</b></br>'%(u,v)\n",
    "    annotation += '<b>Weight</b>: %.3f</br>' % (edge_weights[(u,v)])\n",
    "    annotation += '<b>Edge Ranking</b>: %s' % (k)\n",
    "\n",
    "    annotation += '<hr /><h><b>Sources of Evidence</b></h>'\n",
    "    annotation += evidenceToHTML(u,v,evidence[(u,v)])\n",
    "\n",
    "    return annotation\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Dictionaries of node and edge properties\n",
    "NODE_COLORS = {\n",
    "        #'target'  : '#FFFF60',  # yellow\n",
    "        #'source'  : '#8CE1DC',  # light blue\n",
    "        'target'  : '#4286f4',  # blue\n",
    "        'source'  : '#4286f4',  # blue\n",
    "        'default' : '#D8D8D8',  # gray\n",
    "        'kegg'    : '#ad6cfc',  # purple\n",
    "        'netpath' : '#4286f4',  # blue\n",
    "}\n",
    "NODE_SHAPES = {\n",
    "        'source'  : 'triangle',\n",
    "        'target'  : 'rectangle',\n",
    "        'default' : 'ellipse',\n",
    "}\n",
    "EDGE_COLORS = {\n",
    "        'physical'        : '#27AF47',  # green\n",
    "        'phosphorylation' : '#F07406',  # orange\n",
    "        #'enzymatic'       : '#2A69DC',  # blue\n",
    "        'enzymatic'       : '#DD4B4B',  # red\n",
    "        'activation'      : 'grey',\n",
    "        'inhibition'      : 'grey',\n",
    "        'spike_regulation': 'brown',\n",
    "        'kegg'            : '#ad6cfc', # purple\n",
    "        'netpath'         : '#4286f4',  # blue\n",
    "}\n",
    "\n",
    "# Adds all of the nodes and edges to the GraphSpace object with the specified shapes and colors\n",
    "def constructGraph(ranked_edges, sources, targets):\n",
    "    '''\n",
    "    Posts the pathlinker result to graphspace\n",
    "\n",
    "    :param sources: list of source nodes\n",
    "    :param targets: list of target nodes\n",
    "    '''\n",
    "    # NetworkX object\n",
    "    #G = nx.DiGraph(directed=True)\n",
    "    G = GSGraph()\n",
    "\n",
    "\n",
    "    evidence, edge_types, edge_dir = getEvidence(ranked_edges.keys(), evidence_file)\n",
    "\n",
    "    nodes = set([n for u,v in ranked_edges for n in (u,v)])\n",
    "\n",
    "    ## add GraphSpace/Cytoscape.js attributes to all nodes.\n",
    "    for n in nodes:\n",
    "        #default is gray circle\n",
    "        node_type = 'default'\n",
    "        if n in sources:\n",
    "            # if n is the source, make it a triangle\n",
    "            node_type = 'source'\n",
    "        elif n in targets:\n",
    "            # if n is a taret, make it a square\n",
    "            node_type = 'target'\n",
    "\n",
    "        edgeswithnode = set([(t,h) for t,h in ranked_edges if t==n or h==n])\n",
    "        pathswithnode = set([int(ranked_edges[e]) for e in edgeswithnode])\n",
    "        k_value = min(pathswithnode)\n",
    "        # set the name of the node to be the gene name and add the k to the label\n",
    "        gene_name = uniprot_to_gene[n]\n",
    "        node_popup = getNodePopup(n, pathswithnode)\n",
    "        label=\"%s\\n%d\"%(gene_name,k_value)\n",
    "\n",
    "        G.add_node(gene_name, popup=node_popup, label=label, k=k_value)\n",
    "\n",
    "        # now add the style for the node\n",
    "        shape = NODE_SHAPES[node_type]\n",
    "        color = NODE_COLORS[node_type]\n",
    "        if n in netpath_wnt_nodes:\n",
    "            color = NODE_COLORS['netpath']\n",
    "        elif n in kegg_wnt_nodes:\n",
    "            color = NODE_COLORS['kegg']\n",
    "\n",
    "        G.add_node_style(gene_name, shape=shape, color=color, width=45, height=45,\n",
    "                         style='solid', border_color=color, border_width=2, bubble=color)\n",
    "\n",
    "    # Add all of the edges and their Graphspace/Cytoscape.js attributes\n",
    "    for (u,v) in ranked_edges:\n",
    "        # get the main edge type\n",
    "        main_edge_type = getMainEdgeType(u,v,edge_types[(u,v)])\n",
    "\n",
    "        gene_name_u = uniprot_to_gene[u]\n",
    "        gene_name_v = uniprot_to_gene[v]\n",
    "        edge_popup = getEdgePopup(u,v,k_value, evidence)\n",
    "        k_value = ranked_edges[(u,v)]\n",
    "\n",
    "        G.add_edge(gene_name_u,gene_name_v,directed=edge_dir[(u,v)],popup=edge_popup,k=k_value)\n",
    "\n",
    "        # some attributes (such as opacity) are not implemented in graphspace-python. \n",
    "        # For those, we need to make our own dictionary of the attribute and value\n",
    "        attr_dict = {}\n",
    "        #attr_dict['opacity'] = 0.8\n",
    "        color = EDGE_COLORS[main_edge_type]\n",
    "        if (u,v) in netpath_wnt_edges:\n",
    "            color = EDGE_COLORS['netpath']\n",
    "        elif (u,v) in kegg_wnt_edges:\n",
    "            color = EDGE_COLORS['kegg']\n",
    "        # use the edge weight to set the width of the edge\n",
    "        width = edge_weights[(u,v)] * 2\n",
    "        if 'activation' not in edge_types[(u,v)] and 'inhibition' in edge_types[(u,v)]:\n",
    "            arrow_shape = 'tee'\n",
    "        else:\n",
    "            arrow_shape = 'triangle'\n",
    "\n",
    "        G.add_edge_style(gene_name_u, gene_name_v, attr_dict=attr_dict,\n",
    "                         directed=edge_dir[(u,v)], color=color, width=1.5, arrow_shape=arrow_shape,edge_style='solid')\n",
    "    return G\n",
    "\n",
    "\n",
    "def getMainEdgeType(u,v, edge_types):\n",
    "    \"\"\" a single edge can have multiple edge types according to the different sources or databases\n",
    "    Choose a main edge type here\n",
    "    *edge_types* the set of edge types for a given edge \n",
    "    \"\"\"\n",
    "    main_edge_type = None\n",
    "    edge_type_order = ['phosphorylation', 'enzymatic', 'spike_regulation', 'activation', 'inhibition', 'physical']\n",
    "    for edge_type in edge_type_order:\n",
    "        if edge_type in edge_types:\n",
    "            main_edge_type = edge_type \n",
    "            break\n",
    "            \n",
    "    if main_edge_type is None:\n",
    "        print(\"Warning: edge type of %s->%s not found. Setting to activation\" % (u,v))\n",
    "        main_edge_type = 'activation'\n",
    "\n",
    "    return main_edge_type\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading evidence file data/2017_06-human-interactome-evidence.tsv\n",
      "Warning: edge type of O75197->P19022 not found. Setting to activation\n",
      "Warning: edge type of O75581->Q03135 not found. Setting to activation\n",
      "\n",
      "Graph 'netpath-wnt-pathlinker-k100' already exists. Updating it\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from graphspace_python.api.client import GraphSpace\n",
    "from graphspace_python.graphs.classes.gsgraph import GSGraph\n",
    "import networkx as nx\n",
    "\n",
    "# now finally construct the GraphSpace graph and post it to GraphSpace!\n",
    "G = constructGraph(ranked_edges, sources, targets)\n",
    "\n",
    "# TODO add a description of the graph with a legend describing shapes and colors\n",
    "desc = ''\n",
    "tags = ['pathlinker', 'netpath', '2017-icsb-tutorial']\n",
    "title = 'NetPath WNT Pathway PathLinker Reconstruction'\n",
    "metadata = {'description':desc,'tags':tags, 'title':title}\n",
    "G.set_data(metadata)\n",
    "\n",
    "graph_name = \"netpath-wnt-pathlinker-k100\"\n",
    "G.set_name(graph_name)\n",
    "    \n",
    "# post to graphspace\n",
    "gs = GraphSpace('user6@example.com', 'user6')\n",
    "gs_graph = gs.get_graph(graph_name, owner_email='user6@example.com')\n",
    "if gs_graph is None:\n",
    "    print(\"\\nPosting graph '%s' to graphspace\\n\" % (graph_name))\n",
    "    gs_graph = gs.post_graph(G)\n",
    "else:\n",
    "    print(\"\\nGraph '%s' already exists. Updating it\\n\" % (graph_name))\n",
    "    gs_graph = gs.update_graph(graph_name, owner_email='user6@example.com', graph=G)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reconstruction of the WNT Signaling Pathway\n",
    "![NetPath-WNT-PathLinker-k100](images/NetPath-WNT-PathLinker-k100.png)\n",
    "\n",
    "- First non-WNT pathay node is CFTR\n",
    "  - Experimentally validated RYK-CFTR-DAB2 WNT signal\n",
    "  - Ritz et al., Pathways on demand: automated reconstruction of human signaling networks. npj: Systems Biology and Applications, 2016."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Post your own networks!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
